ballerina-shell/modules/shell-cli/src/main/java/io/ballerina/shell/cli/BallerinaShell.java

```diff
@@ -0,0 +1,201 @@
+        this.configuration = configuration;
+        this.terminal = terminal;
+        this.isRunning = true;
+        this.evaluator = configuration.getEvaluator();
+        this.commandHandler = createCommandHandler();
+    }
+
+    /**
+     * Runs the terminal application using the given config.
+     */
+    public void run() {
+        String leftPrompt = terminal.color(PropertiesLoader.getProperty(REPL_PROMPT),
+                TerminalAdapter.GREEN);
+        terminal.println(FileUtils.readResource(PropertiesLoader.getProperty(HEADER_FILE)));
+
+        Instant start = Instant.now();
+        // Initialize. This must not fail.
+        // If this fails, an error would be directly thrown.
+        try {
+            evaluator.initialize();
+        } catch (BallerinaShellException e) {
+            evaluator.diagnostics().forEach(this::outputDiagnostic);
+            terminal.println("\nShell Initialization Failed!!!");
+            return;
+        }
+
+        Instant end = Instant.now();
+        // Output a warning if initialization took too long.
+        if (Duration.between(start, end).getSeconds() > INIT_WARNING_THRESH_S) {
+            terminal.warn("Compiler initialization took longer than expected.");
+        }
+
+        while (isRunning) {
+            Duration previousDuration = Duration.between(start, end);
+            String rightPrompt = String.format("took %s ms", previousDuration.toMillis());
+            rightPrompt = terminal.color(rightPrompt, TerminalAdapter.BRIGHT);
+
+            String source = terminal.readLine(leftPrompt, rightPrompt).trim();
+
+            start = Instant.now();
+            try {
+                if (!commandHandler.handle(source)) {
+                    String result = evaluator.evaluate(source);
+                    terminal.result(result);
+                }
+            } catch (Exception e) {
+                if (!evaluator.hasErrors()) {
+                    terminal.fatalError("Something went wrong: " + e.getMessage());
+                }
+                outputException(e);
+            } finally {
+                end = Instant.now();
+                evaluator.diagnostics().forEach(this::outputDiagnostic);
+                evaluator.resetDiagnostics();
+                terminal.println("");
+            }
+        }
+    }
+
+    /**
+     * Output a diagnostic to the terminal.
+     *
+     * @param diagnostic Diagnostic to output.
+     */
+    protected void outputDiagnostic(Diagnostic diagnostic) {
+        if (diagnostic.getKind() == DiagnosticKind.DEBUG &&
+                !configuration.isDebug()) {
+            return;
+        }
+
+        if (diagnostic.getKind() == DiagnosticKind.ERROR) {
+            terminal.error(diagnostic.toString());
+        } else if (diagnostic.getKind() == DiagnosticKind.WARN) {
+            terminal.warn(diagnostic.toString());
+        } else {
+            terminal.debug(diagnostic.toString());
+        }
```

Comment:
We can extract `diagnostic.getKind()` out to a variable.

Also, we usually try to return early. e.g,
```suggestion
        DiagnosticKind kind = diagnostic.getKind();
        if (kind == DiagnosticKind.DEBUG) {
            if (!configuration.isDebug()) {
                return;
            }

            terminal.debug(diagnostic.toString());
            return;
        }

        if (kind == DiagnosticKind.ERROR) {
            terminal.error(diagnostic.toString());
            return;
        }
        
        terminal.warn(diagnostic.toString());
```

Reply:
May apply elsewhere too.

Reply:
I refactored the mode out.